src/react/utils.js

```diff
@@ -58,6 +60,95 @@ export function valueIsClassComponent(realm: Realm, value: Value) {
   return false;
 }
 
+// a nested object of a React Element should be hoisted where all its properties are known
+// at evaluation time to be "static" or "pure". This means function values in scope, strings, numbers
+// and deeply nested object/arrays that share the same heuristics
+function canHoistObject(realm: Realm, object: ObjectValue, hoistedScope?: Generator): boolean {
+  if (isReactElement(object)) {
+    return canHoistReactElement(realm, object);
+  }
+  for (let [propName] of object.properties) {
+    let prop = Get(realm, object, propName);
+    if (prop instanceof AbstractValue) {
+      return false;
+    } else if (prop instanceof FunctionValue && prop.parent !== hoistedScope) {
+      return false;
+    } else if (prop instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (prop instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
```

Comment:
Maybe we should be sure that this check is exhaustive since this might miss some things? E.g. check isPrimitive?